### ISSUE TITLE
2.12: case objects are not serializable when extending class

### ISSUE THREAD

**Issue by RustedBones** [scala/bug#13003]
when declaring case objects that extends `class` instead of `trait`, serialization fails with scala 2.12.
case objects are supposed to be serializable.

## Reproduction steps

Scala version: `2.12.19` (also testing on `2.12.x` scala branch)

```scala
package scala

import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4

import java.io.{ByteArrayInputStream, ByteArrayOutputStream, ObjectInputStream, ObjectOutputStream}

sealed abstract class Status(val code: Int)
case object On extends Status(1)
case object Off extends Status(0)

@RunWith(classOf[JUnit4])
class CaseObjectSerializationTest {

  @Test
  def canSerializeMatchError = {
    val barrayOut = new ByteArrayOutputStream()
    new ObjectOutputStream(barrayOut).writeObject(On)
    val barrayIn = new ByteArrayInputStream(barrayOut.toByteArray)
    val status = new ObjectInputStream(barrayIn).readObject().asInstanceOf[Status]
    assert(status == On)
  }
}
```

Added the test from `scala/2.12.x` branch: [diff](https://github.com/scala/scala/compare/2.12.x...RustedBones:scala:case-object-serialization?expand=1)

## Problem

The test fails with
```
[error] Test scala.CaseObjectSerializationTest.canSerializeMatchError failed: java.io.InvalidClassException: scala.On$; no valid constructor, took 0.003 sec
[error]     at java.io.ObjectStreamClass$ExceptionInfo.newInvalidClassException(ObjectStreamClass.java:158)
[error]     at java.io.ObjectStreamClass.checkDeserialize(ObjectStreamClass.java:746)
[error]     at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2202)
[error]     at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1687)
[error]     at java.io.ObjectInputStream.readObject(ObjectInputStream.java:489)
[error]     at java.io.ObjectInputStream.readObject(ObjectInputStream.java:447)
[error]     at scala.CaseObjectSerializationTest.canSerializeMatchError(CaseObjectSerializationTest.scala:21)
[error]     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error]     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error]     at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error]     at java.lang.reflect.Method.invoke(Method.java:566)
[error]     ...
```

**Comment by RustedBones** [comment]
Did some digging and found the cause of the issue / resolution in scala 2.13 [here](https://github.com/scala/scala/commit/2ef533f0423a42ca1d0260d97ba65f2367a65fa6)

- scala 2.12: case object have synthetic method `readResolve`. This replaces the de-serialized object with the singleton instance. However, the de-serialization of the original object fails before the replacement. Indeed, a Serializable class must have access to the no-arg constructor of its first non-serializable superclass.

- scala 2.13: case object have synthetic method `writeReplace` and uses a proxy to de-serialize the object to return the singleton instance. No failure here.

**Comment by SethTisue** [comment]
right... fundamentally a duplicate of #10412 , I think? though the failure mode is different. in any case, closing since it's fixed as of 2.13.0 and we don't keep 2.12-only tickets open
